ic-agent/src/agent/mod.rs

```diff
@@ -891,6 +976,39 @@ impl Agent {
             signed_request_status,
         })
     }
+
+    async fn get_subnet_by_canister(
+        &self,
+        canister: &Principal,
+    ) -> Result<Arc<Subnet>, AgentError> {
+        let subnet = self
+            .subnet_key_cache
+            .lock()
+            .unwrap()
+            .get_subnet_by_canister(canister);
+        if let Some(subnet) = subnet {
+            Ok(subnet)
+        } else {
+            let cert = self
+                .read_state_raw(vec![vec!["subnet".into()]], *canister)
+                .await?;
+            let time = leb128::read::unsigned(&mut lookup_value(&cert.tree, [b"time".as_ref()])?)?;
+            if (OffsetDateTime::now_utc()
+                - OffsetDateTime::from_unix_timestamp_nanos(time as _).unwrap())
```

Comment:
Can we return an error rather than panic here?

Reply:
See other comment.